### PR TITLE
add rosdistro to auto tick bot

### PR DESCRIPTION
**Pull request by wolfv**
First iteration of auto tick bot that supports ROSDistro.

### CODE REVIEW

File: conda_forge_tick/update_upstream_versions.py [hunk 1]
```diff
@@ -366,7 +435,7 @@ def _update_upstream_versions_process_pool(gx, sources):
 
 def update_upstream_versions(gx, sources=None):
     sources = (
-        (PyPI(), NPM(), CRAN(), RawURL(), Github()) if sources is None else sources
+        (PyPI(), NPM(), CRAN(), RawURL(), Github(), ROSDistro()) if sources is None else sources
```
Comment (CJ-Wright) [inline conda_forge_tick/update_upstream_versions.py]: You might want to put the `ROSDistro` before `RawURL`

**Comment by CJ-Wright** [comment]
The travis error I think is unrelated to this PR. I think I fixed it on master, but we'll see.

**Comment by wolfv** [comment]
Hey, yes I tested it and it seemed to work very well. It should also only apply to the `ros-` subset of packages :) I’ll move it in front of RAWUrl()!

**Comment by wolfv** [comment]
Yeah, I tested locally. This is teh output I get:

```
2019-10-28 15:47:16,818 INFO     conda_forge_tick.update_upstream_versions || Reading graph
2019-10-28 15:47:17,081 INFO     conda_forge_tick.update_upstream_versions || Updating upstream versions
2019-10-28 15:47:22,443 INFO     conda_forge_tick.update_upstream_versions || ROS Distro source initialized
/home/wolfv/Programs/cf-scripts/conda_forge_tick/update_upstream_versions.py:402: DeprecationWarning: The 'warn' method is deprecated, use 'warning' instead
  logger.warn("Error getting uptream version of {}: {}".format(node, se))
2019-10-28 15:47:22,444 WARNING  conda_forge_tick.update_upstream_versions || Error getting uptream version of ros-conda-base: 'ros-conda-base'
2019-10-28 15:47:22,449 WARNING  conda_forge_tick.update_upstream_versions || Error getting uptream version of ros-conda-mutex: 'ros-conda-mutex'
/home/wolfv/Programs/cf-scripts/conda_forge_tick/update_upstream_versions.py:273: DeprecationWarning: The 'warn' method is deprecated, use 'warning' instead
  logger.warn(version)
2019-10-28 15:47:22,450 WARNING  conda_forge_tick.update_upstream_versions || 0.7.19-1
2019-10-28 15:47:22,450 INFO     conda_forge_tick.update_upstream_versions || ros-catkin - 0.7.17 - 0.7.19
2019-10-28 15:47:22,451 WARNING  conda_forge_tick.update_upstream_versions || 1.2.2-1
2019-10-28 15:47:22,452 INFO     conda_forge_tick.update_upstream_versions || ros-environment - 1.2.1 - 1.2.2
2019-10-28 15:47:22,452 WARNING  conda_forge_tick.update_upstream_versions || 0.6.13-1
2019-10-28 15:47:22,453 INFO     conda_forge_tick.update_upstream_versions || ros-cpp-common - 0.6.12 - 0.6.13
2019-10-28 15:47:22,454 WARNING  conda_forge_tick.update_upstream_versions || 0.5.12-0
2019-10-28 15:47:22,454 INFO     conda_forge_tick.update_upstream_versions || ros-genmsg - 0.5.12 - 0.5.12
2019-10-28 15:47:22,455 WARNING  conda_forge_tick.update_upstream_versions || 0.6.13-1
2019-10-28 15:47:22,456 INFO     conda_forge_tick.update_upstream_versions || ros-rostime - 0.6.12 - 0.6.13
2019-10-28 15:47:22,457 WARNING  conda_forge_tick.update_upstream_versions || 0.4.2-0
2019-10-28 15:47:22,458 INFO     conda_forge_tick.update_upstream_versions || ros-cmake-modules - 0.4.2 - 0.4.2
2019-10-28 15:47:22,459 WARNING  conda_forge_tick.update_upstream_versions || 0.6.2-0
2019-10-28 15:47:22,460 INFO     conda_forge_tick.update_upstream_versions || ros-gencpp - 0.6.2 - 0.6.2
2019-10-28 15:47:22,460 WARNING  conda_forge_tick.update_upstream_versions || 0.6.13-1
2019-10-28 15:47:22,461 INFO     conda_forge_tick.update_upstream_versions || ros-roscpp-traits - 0.6.12 - 0.6.13
2019-10-28 15:47:22,462 WARNING  conda_forge_tick.update_upstream_versions || 2.0.1-0
2019-10-28 15:47:22,462 INFO     conda_forge_tick.update_upstream_versions || ros-gennodejs - 2.0.1 - 2.0.1
2019-10-28 15:47:22,463 WARNING  conda_forge_tick.update_upstream_versions || 2.5.4-1
2019-10-28 15:47:22,463 INFO     conda_forge_tick.update_upstream_versions || ros-rospack - 2.5.3 - 2.5.4
2019-10-28 15:47:22,464 WARNING  conda_forge_tick.update_upstream_versions || 0.4.0-0
2019-10-28 15:47:22,465 INFO     conda_forge_tick.update_upstream_versions || ros-message-generation - 0.4.0 - 0.4.0
2019-10-28 15:47:22,465 WARNING  conda_forge_tick.update_upstream_versions || 0.4.1-0
2019-10-28 15:47:22,466 INFO     conda_forge_tick.update_upstream_versions || ros-class-loader - 0.4.1 - 0.4.1
2019-10-28 15:47:22,466 WARNING  conda_forge_tick.update_upstream_versions || 1.14.7-1
2019-10-28 15:47:22,467 INFO     conda_forge_tick.update_upstream_versions || ros-roslib - 1.14.6 - 1.14.7
2019-10-28 15:47:22,468 WARNING  conda_forge_tick.update_upstream_versions || 2.2.6-0
2019-10-28 15:47:22,468 INFO     conda_forge_tick.update_upstream_versions || ros-geneus - 2.2.6 - 2.2.6
2019-10-28 15:47:22,469 WARNING  conda_forge_tick.update_upstream_versions || 0.6.9-1
2019-10-28 15:47:22,470 INFO     conda_forge_tick.update_upstream_versions || ros-genpy - 0.6.8 - 0.6.9
2019-10-28 15:47:22,470 WARNING  conda_forge_tick.update_upstream_versions || 0.4.16-0
2019-10-28 15:47:22,471 INFO     conda_forge_tick.update_upstream_versions || ros-genlisp - 0.4.16 - 0.4.16
2019-10-28 15:47:22,472 WARNING  conda_forge_tick.update_upstream_versions || 1.14.7-1
2019-10-28 15:47:22,473 INFO     conda_forge_tick.update_upstream_versions || ros-rosmake - 1.14.6 - 1.14.7
2019-10-28 15:47:22,473 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,474 INFO     conda_forge_tick.update_upstream_versions || ros-rosgraph - 1.14.3 - 1.14.3
2019-10-28 15:47:22,475 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,476 INFO     conda_forge_tick.update_upstream_versions || ros-xmlrpcpp - 1.14.3 - 1.14.3
2019-10-28 15:47:22,477 WARNING  conda_forge_tick.update_upstream_versions || 0.4.12-0
2019-10-28 15:47:22,478 INFO     conda_forge_tick.update_upstream_versions || ros-message-runtime - 0.4.12 - 0.4.12
2019-10-28 15:47:22,478 WARNING  conda_forge_tick.update_upstream_versions || 0.6.13-1
2019-10-28 15:47:22,479 INFO     conda_forge_tick.update_upstream_versions || ros-roscpp-serialization - 0.6.12 - 0.6.13
2019-10-28 15:47:22,480 WARNING  conda_forge_tick.update_upstream_versions || 0.6.13-1
2019-10-28 15:47:22,480 INFO     conda_forge_tick.update_upstream_versions || ros-roscpp-core - 0.6.12 - 0.6.13
2019-10-28 15:47:22,481 WARNING  conda_forge_tick.update_upstream_versions || 1.14.7-1
2019-10-28 15:47:22,481 INFO     conda_forge_tick.update_upstream_versions || ros-roscreate - 1.14.6 - 1.14.7
2019-10-28 15:47:22,482 WARNING  conda_forge_tick.update_upstream_versions || 1.14.7-1
2019-10-28 15:47:22,483 INFO     conda_forge_tick.update_upstream_versions || ros-rosboost-cfg - 1.14.6 - 1.14.7
2019-10-28 15:47:22,483 WARNING  conda_forge_tick.update_upstream_versions || 1.14.7-1
2019-10-28 15:47:22,484 INFO     conda_forge_tick.update_upstream_versions || ros-rosunit - 1.14.6 - 1.14.7
2019-10-28 15:47:22,484 WARNING  conda_forge_tick.update_upstream_versions || 1.14.7-1
2019-10-28 15:47:22,485 INFO     conda_forge_tick.update_upstream_versions || ros-rosbash - 1.14.6 - 1.14.7
2019-10-28 15:47:22,485 WARNING  conda_forge_tick.update_upstream_versions || 0.5.12-0
2019-10-28 15:47:22,486 INFO     conda_forge_tick.update_upstream_versions || ros-std-msgs - 0.5.12 - 0.5.12
2019-10-28 15:47:22,487 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,487 INFO     conda_forge_tick.update_upstream_versions || ros-rosmaster - 1.14.3 - 1.14.3
2019-10-28 15:47:22,488 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,489 INFO     conda_forge_tick.update_upstream_versions || ros-rosparam - 1.14.3 - 1.14.3
2019-10-28 15:47:22,490 WARNING  conda_forge_tick.update_upstream_versions || 1.11.2-0
2019-10-28 15:47:22,491 INFO     conda_forge_tick.update_upstream_versions || ros-std-srvs - 1.11.2 - 1.11.2
2019-10-28 15:47:22,491 WARNING  conda_forge_tick.update_upstream_versions || 1.12.7-0
2019-10-28 15:47:22,492 INFO     conda_forge_tick.update_upstream_versions || ros-diagnostic-msgs - 1.12.7 - 1.12.7
2019-10-28 15:47:22,493 WARNING  conda_forge_tick.update_upstream_versions || 1.12.7-0
2019-10-28 15:47:22,493 INFO     conda_forge_tick.update_upstream_versions || ros-actionlib-msgs - 1.12.7 - 1.12.7
2019-10-28 15:47:22,494 WARNING  conda_forge_tick.update_upstream_versions || 1.12.7-0
2019-10-28 15:47:22,495 INFO     conda_forge_tick.update_upstream_versions || ros-geometry-msgs - 1.12.7 - 1.12.7
2019-10-28 15:47:22,495 WARNING  conda_forge_tick.update_upstream_versions || 1.14.7-1
2019-10-28 15:47:22,496 INFO     conda_forge_tick.update_upstream_versions || ros-roslang - 1.14.6 - 1.14.7
2019-10-28 15:47:22,496 WARNING  conda_forge_tick.update_upstream_versions || 1.11.2-0
2019-10-28 15:47:22,497 INFO     conda_forge_tick.update_upstream_versions || ros-rosgraph-msgs - 1.11.2 - 1.11.2
2019-10-28 15:47:22,498 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,498 INFO     conda_forge_tick.update_upstream_versions || ros-roslz4 - 1.14.3 - 1.14.3
2019-10-28 15:47:22,499 WARNING  conda_forge_tick.update_upstream_versions || 1.0.0-0
2019-10-28 15:47:22,500 INFO     conda_forge_tick.update_upstream_versions || ros-rosbag-migration-rule - 1.0.0 - 1.0.0
2019-10-28 15:47:22,500 WARNING  conda_forge_tick.update_upstream_versions || 1.14.7-1
2019-10-28 15:47:22,501 INFO     conda_forge_tick.update_upstream_versions || ros-rosbuild - 1.14.6 - 1.14.7
2019-10-28 15:47:22,502 WARNING  conda_forge_tick.update_upstream_versions || 1.12.7-0
2019-10-28 15:47:22,503 INFO     conda_forge_tick.update_upstream_versions || ros-shape-msgs - 1.12.7 - 1.12.7
2019-10-28 15:47:22,503 WARNING  conda_forge_tick.update_upstream_versions || 1.12.7-0
2019-10-28 15:47:22,504 INFO     conda_forge_tick.update_upstream_versions || ros-trajectory-msgs - 1.12.7 - 1.12.7
2019-10-28 15:47:22,505 WARNING  conda_forge_tick.update_upstream_versions || 1.12.7-0
2019-10-28 15:47:22,506 INFO     conda_forge_tick.update_upstream_versions || ros-visualization-msgs - 1.12.7 - 1.12.7
2019-10-28 15:47:22,507 WARNING  conda_forge_tick.update_upstream_versions || 1.12.7-0
2019-10-28 15:47:22,508 INFO     conda_forge_tick.update_upstream_versions || ros-sensor-msgs - 1.12.7 - 1.12.7
2019-10-28 15:47:22,509 WARNING  conda_forge_tick.update_upstream_versions || 1.12.7-0
2019-10-28 15:47:22,510 INFO     conda_forge_tick.update_upstream_versions || ros-stereo-msgs - 1.12.7 - 1.12.7
2019-10-28 15:47:22,511 WARNING  conda_forge_tick.update_upstream_versions || 1.9.22-0
2019-10-28 15:47:22,512 INFO     conda_forge_tick.update_upstream_versions || ros-roslisp - 1.9.22 - 1.9.22
2019-10-28 15:47:22,512 WARNING  conda_forge_tick.update_upstream_versions || 1.13.11-1
2019-10-28 15:47:22,513 INFO     conda_forge_tick.update_upstream_versions || ros-rosconsole - 1.13.10 - 1.13.11
2019-10-28 15:47:22,514 WARNING  conda_forge_tick.update_upstream_versions || 1.12.7-0
2019-10-28 15:47:22,514 INFO     conda_forge_tick.update_upstream_versions || ros-common-msgs - 1.12.7 - 1.12.7
2019-10-28 15:47:22,515 WARNING  conda_forge_tick.update_upstream_versions || 1.12.7-0
2019-10-28 15:47:22,516 INFO     conda_forge_tick.update_upstream_versions || ros-nav-msgs - 1.12.7 - 1.12.7
2019-10-28 15:47:22,516 WARNING  conda_forge_tick.update_upstream_versions || 1.12.1-0
2019-10-28 15:47:22,517 INFO     conda_forge_tick.update_upstream_versions || ros-pluginlib - 1.12.1 - 1.12.1
2019-10-28 15:47:22,518 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,519 INFO     conda_forge_tick.update_upstream_versions || ros-roscpp - 1.14.3 - 1.14.3
2019-10-28 15:47:22,519 WARNING  conda_forge_tick.update_upstream_versions || 1.14.7-1
2019-10-28 15:47:22,520 INFO     conda_forge_tick.update_upstream_versions || ros-mk - 1.14.6 - 1.14.7
2019-10-28 15:47:22,520 WARNING  conda_forge_tick.update_upstream_versions || 1.14.7-1
2019-10-28 15:47:22,521 INFO     conda_forge_tick.update_upstream_versions || ros-ros - 1.14.6 - 1.14.7
2019-10-28 15:47:22,522 WARNING  conda_forge_tick.update_upstream_versions || 1.14.7-1
2019-10-28 15:47:22,522 INFO     conda_forge_tick.update_upstream_versions || ros-rosclean - 1.14.6 - 1.14.7
2019-10-28 15:47:22,523 WARNING  conda_forge_tick.update_upstream_versions || 0.5.3-0
2019-10-28 15:47:22,523 INFO     conda_forge_tick.update_upstream_versions || ros-rosconsole-bridge - 0.5.3 - 0.5.3
2019-10-28 15:47:22,524 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,525 INFO     conda_forge_tick.update_upstream_versions || ros-rospy - 1.14.3 - 1.14.3
2019-10-28 15:47:22,525 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,526 INFO     conda_forge_tick.update_upstream_versions || ros-rosout - 1.14.3 - 1.14.3
2019-10-28 15:47:22,527 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,527 INFO     conda_forge_tick.update_upstream_versions || ros-roslaunch - 1.14.3 - 1.14.3
2019-10-28 15:47:22,528 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,529 INFO     conda_forge_tick.update_upstream_versions || ros-rostest - 1.14.3 - 1.14.3
2019-10-28 15:47:22,529 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,530 INFO     conda_forge_tick.update_upstream_versions || ros-rosbag-storage - 1.14.3 - 1.14.3
2019-10-28 15:47:22,531 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,532 INFO     conda_forge_tick.update_upstream_versions || ros-message-filters - 1.14.3 - 1.14.3
2019-10-28 15:47:22,532 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,533 INFO     conda_forge_tick.update_upstream_versions || ros-topic-tools - 1.14.3 - 1.14.3
2019-10-28 15:47:22,534 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,535 INFO     conda_forge_tick.update_upstream_versions || ros-rosbag - 1.14.3 - 1.14.3
2019-10-28 15:47:22,536 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,537 INFO     conda_forge_tick.update_upstream_versions || ros-rosmsg - 1.14.3 - 1.14.3
2019-10-28 15:47:22,538 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,539 INFO     conda_forge_tick.update_upstream_versions || ros-rostopic - 1.14.3 - 1.14.3
2019-10-28 15:47:22,540 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,541 INFO     conda_forge_tick.update_upstream_versions || ros-rosnode - 1.14.3 - 1.14.3
2019-10-28 15:47:22,541 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,542 INFO     conda_forge_tick.update_upstream_versions || ros-rosservice - 1.14.3 - 1.14.3
2019-10-28 15:47:22,543 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,544 INFO     conda_forge_tick.update_upstream_versions || ros-roswtf - 1.14.3 - 1.14.3
2019-10-28 15:47:22,544 WARNING  conda_forge_tick.update_upstream_versions || 1.12.0-1
2019-10-28 15:47:22,545 INFO     conda_forge_tick.update_upstream_versions || ros-actionlib - 1.11.15 - 1.12.0
2019-10-28 15:47:22,546 WARNING  conda_forge_tick.update_upstream_versions || 1.14.3-0
2019-10-28 15:47:22,547 INFO     conda_forge_tick.update_upstream_versions || ros-comm - 1.14.3 - 1.14.3
2019-10-28 15:47:22,547 WARNING  conda_forge_tick.update_upstream_versions || 1.4.1-0
2019-10-28 15:47:22,548 INFO     conda_forge_tick.update_upstream_versions || ros-core - 1.4.1 - 1.4.1
2019-10-28 15:47:22,549 WARNING  conda_forge_tick.update_upstream_versions || 1.6.0-0
2019-10-28 15:47:22,549 INFO     conda_forge_tick.update_upstream_versions || ros-dynamic-reconfigure - 1.6.0 - 1.6.0
2019-10-28 15:47:23,519 INFO     conda_forge_tick.update_upstream_versions || Current number of out of date packages not PRed: 99
2019-10-28 15:47:23,519 INFO     conda_forge_tick.update_upstream_versions || writing out file
```

However, one issue that I am not sure is resolved is the following:

The version is sometimes `0.6.13`, but the package URL is at `.../0.6.13-1.tar.gz`

This URL is correctly returned from `get_url(...)`. But is this URL also used to generate the PR? I didn't find out how to test the PR generation ... 

Would be great if this gets merged asap because I'd love to see the bot working for the ROS distro :)

**Comment by wolfv** [comment]
Hmm, so I’d need to split the version on the jinja side? It’s gonna be a little painful to update all recipes :)

**Comment by CJ-Wright** [comment]
Yeah, it's better to have the complexity on the jinja2 side, since the bot's regex are a bit rough.

**Comment by CJ-Wright** [comment]
You could write a piggy back migrator to fix the urls

**Comment by CJ-Wright** [comment]
Thank you very much for this!

I'm going to merge this and see how it goes. If it causes dask issues (we've seen these before) I may need to revert.

**Pull request merged**